File: algorithms/107.py
# https://leetcode.com/problems/binary-tree-level-order-traversal-ii


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """

        # bfs
        result = []
        queue = [(0, root)]
        while len(queue) > 0:
            level, node = queue.pop(0)
            if node is None:
                continue
            if len(result) < level + 1:
                result.insert(0, [])
            result[-level - 1].append(node.val)
            queue.append((level + 1, node.left))
            queue.append((level + 1, node.right))
        return result

File: algorithms/107.swift
// https://leetcode.com/problems/binary-tree-level-order-traversal-ii

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func levelOrderBottom(root: TreeNode?) -> [[Int]] {
        // dfs
        func trace(root: TreeNode?, _ level: Int, inout _ result: [[Int]]) {
            guard let r = root else { return }
            if result.count < level+1 { result.insert([], atIndex: 0) }
            result[result.count-level-1].append(r.val)
            trace(r.left, level+1, &result)
            trace(r.right, level+1, &result)
        }
        var result = [[Int]]()
        trace(root, 0, &result)
        return result
    }
}
